assertion failed: failure to construct path from value. object foo in class Bar does not have an outer accessor

## Compiler version
3.0.0-RC1

## Minimized code

```Scala
class Wrapper[E <: Throwable]:
  object syntax:
    extension [A <: Matchable](a: E | A) 
      inline def foreach(f: A => Any): Unit = 
        a match
          case e: E => ()
          case a: A => f(a)

  def _catch[A](a: => A): E | A = 
    try a
    catch case e: E => e

object throwables extends Wrapper[Throwable]
import throwables.syntax.*

@main def app(): Unit = 
  for x <- throwables._catch(1/1) do println(x)
```

## Output (click arrow to expand)
<details>

```scala
[info] compiling 1 Scala source to /Users/tamer.abdulradi/work/tabdulradi/failable/core/target/scala-3.0.0-RC1/test-classes ...
[info] exception occurred while compiling /Users/tamer.abdulradi/work/tabdulradi/failable/core/src/test/scala/com/abdulradi/nullable/Examples.scala
java.lang.AssertionError: assertion failed: failure to construct path from value Wrapper_this/method app/package object Examples$package/package test/package <root> to `this` of  <none>;
object syntax in class Wrapper does not have an outer accessor while compiling /Users/tamer.abdulradi/work/tabdulradi/failable/core/src/test/scala/com/abdulradi/nullable/Examples.scala
[error] ## Exception when compiling 1 sources to /Users/tamer.abdulradi/work/tabdulradi/failable/core/target/scala-3.0.0-RC1/test-classes
[error] java.lang.AssertionError: assertion failed: failure to construct path from value Wrapper_this/method app/package object Examples$package/package test/package <root> to `this` of  <none>;
[error] object syntax in class Wrapper does not have an outer accessor
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.loop$1(ExplicitOuter.scala:412)
[error] dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.path$extension(ExplicitOuter.scala:431)
[error] dotty.tools.dotc.transform.ElimOuterSelect.transformSelect(ElimOuterSelect.scala:29)
[error] dotty.tools.dotc.transform.MegaPhase.goSelect(MegaPhase.scala:599)
[error] dotty.tools.dotc.transform.MegaPhase.goSelect(MegaPhase.scala:600)
[error] dotty.tools.dotc.transform.MegaPhase.goSelect(MegaPhase.scala:600)
[error] dotty.tools.dotc.transform.MegaPhase.goSelect(MegaPhase.scala:600)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:229)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
[error] dotty.tools.dotc.transform.MegaPhase.mapValDef$1(MegaPhase.scala:235)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:240)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
[error] dotty.tools.dotc.transform.MegaPhase.recur$2(MegaPhase.scala:453)
[error] dotty.tools.dotc.transform.MegaPhase.recur$2(MegaPhase.scala:453)
[error] dotty.tools.dotc.transform.MegaPhase.recur$2(MegaPhase.scala:453)
[error] dotty.tools.dotc.transform.MegaPhase.recur$2(MegaPhase.scala:453)
[error] dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:453)
[error] dotty.tools.dotc.transform.MegaPhase.transformSpecificTrees(MegaPhase.scala:456)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:396)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
[error] dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:249)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:252)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
[error] dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:437)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:362)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
[error] dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:256)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
[error] dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:437)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.recur$1(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:442)
[error] dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:382)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:385)
[error] dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:429)
[error] dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:448)
[error] dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:460)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:230)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] dotty.tools.dotc.Run.compile(Run.scala:150)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2200)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2157)
[error] sbt.internal.io.Retry$.apply(Retry.scala:40)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2153)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:748)
[error]            
[error] stack trace is suppressed; run last core / Test / compileIncremental for the full output
[error] (core / Test / compileIncremental) java.lang.AssertionError: assertion failed: failure to construct path from value Wrapper_this/method app/package object Examples$package/package test/package <root> to `this` of  <none>;
[error] object syntax in class Wrapper does not have an outer accessor
[error] Total time: 1 s, completed 26-Mar-2021 21:59:42
```
</details>

--- a/compiler/src/dotty/tools/dotc/typer/Inliner.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Inliner.scala
@@ -508,14 +508,9 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
   private def computeThisBindings() = {
     // All needed this-proxies, paired-with and sorted-by nesting depth of
     // the classes they represent (innermost first)
-    val sortedProxies = thisProxy.toList.map {
-      case (cls, proxy) =>
-        // The class that the this-proxy `selfSym` represents
-        def classOf(selfSym: Symbol) = selfSym.info.classSymbol
-        // The total nesting depth of the class represented by `selfSym`.
-        def outerLevel(selfSym: Symbol): Int = classOf(selfSym).ownersIterator.length
-        (outerLevel(cls), proxy.symbol)
-    }.sortBy(-_._1)
+    val sortedProxies = thisProxy.toList
+      .map((cls, proxy) => (cls.ownersIterator.length, proxy.symbol))
+      .sortBy(-_._1)
 
     var lastSelf: Symbol = NoSymbol
     var lastLevel: Int = 0

--- a/compiler/src/dotty/tools/dotc/typer/Inliner.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Inliner.scala
@@ -522,9 +522,13 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
     for ((level, selfSym) <- sortedProxies) {
       lazy val rhsClsSym = selfSym.info.widenDealias.classSymbol
       val rhs =
-        if (lastSelf.exists)
-          ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
-        else if (rhsClsSym.is(Module) && rhsClsSym.isStatic)
+        if lastSelf.exists then
+          selfSym.info match
+            case info: TermRef if info.isStable =>
+              ref(info)
+            case _ =>
+              ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
+        else if rhsClsSym.is(Module) && rhsClsSym.isStatic then
           ref(rhsClsSym.sourceModule)
         else
           inlineCallPrefix

--- a/compiler/src/dotty/tools/dotc/typer/Inliner.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Inliner.scala
@@ -516,17 +516,17 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
     var lastLevel: Int = 0
     for ((level, selfSym) <- sortedProxies) {
       lazy val rhsClsSym = selfSym.info.widenDealias.classSymbol
-      val rhs =
-        if lastSelf.exists then
-          selfSym.info match
-            case info: TermRef if info.isStable =>
-              ref(info)
-            case _ =>
-              ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
-        else if rhsClsSym.is(Module) && rhsClsSym.isStatic then
-          ref(rhsClsSym.sourceModule)
-        else
-          inlineCallPrefix
+      val rhs = selfSym.info.dealias match
+        case info: TermRef if info.isStable =>
+          ref(info)
+        case info =>
+          val rhsClsSym = info.widenDealias.classSymbol
+          if rhsClsSym.is(Module) && rhsClsSym.isStatic then
+            ref(rhsClsSym.sourceModule)
+          else if lastSelf.exists then
+            ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
+          else
+            inlineCallPrefix
       val binding = ValDef(selfSym.asTerm, QuoteUtils.changeOwnerOfTree(rhs, selfSym)).withSpan(selfSym.span)
       bindingsBuf += binding
       inlining.println(i"proxy at $level: $selfSym = ${bindingsBuf.last}")

--- a/tests/run/i11914.scala
+++ b/tests/run/i11914.scala
@@ -1,0 +1,17 @@
+class Wrapper[E <: Throwable]:
+  object syntax:
+    extension [A <: Matchable](a: E | A)
+      transparent inline def foreach(f: A => Any): Unit =
+        a match
+          case e: E => ()
+          case a: A => f(a)
+
+  def _catch[A](a: => A): E | A =
+    try a
+    catch case e: E => e
+
+object throwables extends Wrapper[Throwable]
+import throwables.syntax.*
+
+@main def Test(): Unit =
+  for x <- throwables._catch(1/1) do println(x)

--- a/tests/run/i11914a.scala
+++ b/tests/run/i11914a.scala
@@ -1,0 +1,23 @@
+object Wrapper:
+  type E <: Throwable
+  class C:
+    object syntax:
+        extension [A <: Matchable](a: E | A)
+          transparent inline def foreach(f: A => Any): Unit =
+            a match
+              case e: E =>
+                Wrapper.this.n + m
+              case a: A => f(a)
+        val m: Int = 4
+
+  def _catch[A](a: => A): E | A =
+    try a
+    catch case e: E => e
+
+  val n: Int = 3
+
+object throwables extends Wrapper.C
+import throwables.syntax.*
+
+@main def Test(): Unit =
+  for x <- Wrapper._catch(1/1) do println(x)

